// varang/components/core.js
"use strict";
var RemoteProviderPagingParams = (function () {
    function RemoteProviderPagingParams() {
    }
    return RemoteProviderPagingParams;
}());
exports.RemoteProviderPagingParams = RemoteProviderPagingParams;
var RemoteProviderRequestParamNames = (function () {
    function RemoteProviderRequestParamNames() {
    }
    return RemoteProviderRequestParamNames;
}());
exports.RemoteProviderRequestParamNames = RemoteProviderRequestParamNames;
var RemoteDataProviderIn = (function () {
    function RemoteDataProviderIn() {
        this.jasonXPath = new RemoteProviderPagingParams();
    }
    return RemoteDataProviderIn;
}());
exports.RemoteDataProviderIn = RemoteDataProviderIn;
var RemoteDataProvider = (function () {
    function RemoteDataProvider() {
        this.in = new RemoteDataProviderIn();
        this.out = new RemoteProviderRequestParamNames();
    }
    return RemoteDataProvider;
}());
exports.RemoteDataProvider = RemoteDataProvider;
var ClientPagerParams = (function () {
    function ClientPagerParams() {
    }
    return ClientPagerParams;
}());
exports.ClientPagerParams = ClientPagerParams;
var DataSourceProperties = (function () {
    function DataSourceProperties() {
        this.url = "id";
    }
    return DataSourceProperties;
}());
exports.DataSourceProperties = DataSourceProperties;
var VarGridHeaderAction = (function () {
    function VarGridHeaderAction(name, toggle) {
        this.toggle = toggle;
        this.name = name;
        this.enabled = false;
    }
    VarGridHeaderAction.prototype.sord = function () {
        return this.toggle ? "ASC" : "DESC";
    };
    return VarGridHeaderAction;
}());
exports.VarGridHeaderAction = VarGridHeaderAction;
// export interface RemoteProviderPagingParams {
//   list:string;
//   pageSize:string;
//   pageIndex:string;
//   total:string;
// }
// export interface RemoteProviderRequestParamNames {
//   pageStart:string;
//   pageSize:string;
//   sortIndex:string;
//   sortOrder:string;
// }
// export interface RemoteDataProviderIn {
//   jasonXPath:RemoteProviderPagingParams;
// }
// export interface RemoteDataProvider {
//     in:RemoteDataProviderIn;
//     out:RemoteProviderRequestParamNames;
// }
// export interface ClientPagerParams {
//     id:string;
//     sortIndex:string;
//     pageSizes:number[];
//     pageSize:number;
//     sortOrder:string;
//     pageStart:number;
// }
// export class DataSourceProperties {
//   url:string;
// }
